xarray/backends/api.py

```diff
@@ -117,10 +120,20 @@ def check_attr(name, value):
             check_attr(k, v)
 
 
+def _protect_dataset_variables_inplace(dataset, cache):
+    for name, variable in dataset.variables.items():
+        if name not in variable.dims:
+            # no need to protect IndexVariable objects
+            data = indexing.CopyOnWriteArray(variable._data)
```

Comment:
I ran a test where I create a `DataSet` from a custom data store which initializes `Variables` using dask arrays for `data`. In this case the dask arrays is still converted to an ndarray when accessing the `Variable`'s `data` property, since [it checks](https://github.com/shoyer/xarray/blob/49135f20190cd024dbc7148262026659cc36e955/xarray/core/variable.py#L271) is for a dask array type, however here the array is wrapped into a `CopyOnWriteArray`, which means `Variable.values` is called, which loads eagerly.